src/coreclr/src/tools/Common/Compiler/CoreRTNameMangler.cs

```diff
@@ -166,11 +165,14 @@ private string DisambiguateName(string origName, ISet<string> set)
 
         public override string GetMangledTypeName(TypeDesc type)
         {
-            string mangledName;
-            if (_mangledTypeNames.TryGetValue(type, out mangledName))
-                return mangledName;
+            lock (this)
```

Comment:
Can we break down the locking hierarchy so that there's not a single giant lock for type names, method names, field names, and mangled strings, but instead there's a lock for each?

In general, I'm worried that we're probably not using this from multithreaded code at all right now, so this looks like an improvement, but the locks will become a bottleneck as soon as we start doing that.

Reply:
How about using the lock free concurrent dictionaries here, to avoid locking while performing the lookup?

Reply:
@MichalStrehovsky We are using this from multithreaded code at the moment. I tried using more granulated locking and it didn't seem to make a performance difference. @davidwrighton and I thought that, since there's no performance difference right now, it would be better to keep it in this form where it's clear to people reading it that there are no deadlocks.

@fadimounir That's an option, but then it gets a little hard to analyze because it's hard to reason about whether each of the GetOrAdd functions must only be run once and so has to be either locked or wrapped in a Lazy. This + the limited time left is why we didn't end up using that right now, but perhaps @davidwrighton can provide more direction.

Reply:
> because it's hard to reason about whether each of the GetOrAdd functions must only be run once 

What do you mean? We can always do a .TryGetValue outside of a lock, and .TryAdd inside a lock (`Dictionary` has the same APIs too so we can use it as well). Locking while retrieving a value is not ideal

Reply:
Oh I see what you mean now. Yes I could change it to do that, but I'm not sure whether we even have contention there.

Reply:
When using Dictionary you cannot use the TryGetValue apis outside of a lock. This worked before with ImmutableDictionary as that api allows lock-free reads. My expectation is that ConcurrentDictionary would be better in performance here than Dictionary, but I don't want to see us use that without seeing the performance impact in a profiler. I think we should check Anubhav's work in here as written, and explore finding even more performance in the future.